sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -56,33 +58,100 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
+    val primitiveTypeName = if (isPrimitiveArray) ctx.primitiveTypeName(et) else ""
+    val (preprocess, arrayData, arrayWriter) =
+      GenArrayData.getCodeArrayData(ctx, et, children.size, isPrimitiveArray, array)
+
+    val assigns = if (isPrimitiveArray) {
+      evals.zipWithIndex.map { case (eval, i) =>
+        eval.code + s"""
+         if (${eval.isNull}) {
+           $arrayWriter.setNull$primitiveTypeName($i);
+         } else {
+           $arrayWriter.write($i, ${eval.value});
+         }
+       """
+      }
+    } else {
+      evals.zipWithIndex.map { case (eval, i) =>
+        eval.code + s"""
+         if (${eval.isNull}) {
+           $array[$i] = null;
+         } else {
+           $array[$i] = ${eval.value};
+         }
+       """
+      }
+    }
+    ev.copy(code =
+      preprocess +
+      ctx.splitExpressions(ctx.INPUT_ROW, assigns) +
+      s"\nfinal ArrayData ${ev.value} = $arrayData;\n",
+      isNull = "false")
   }
 
   override def prettyName: String = "array"
 }
 
+private [sql] object GenArrayData {
+  // This function returns Java code pieces based on DataType and isPrimitive
+  // for allocation of ArrayData class
+  def getCodeArrayData(
+      ctx: CodegenContext,
+      dt: DataType,
+      size: Int,
+      isPrimitive : Boolean,
+      array: String): (String, String, String) = {
+    if (!isPrimitive) {
+      val arrayClass = classOf[GenericArrayData].getName
+      ctx.addMutableState("Object[]", array,
+        s"this.$array = new Object[${size}];")
+      ("", s"new $arrayClass($array)", null)
+    } else {
+      val row = ctx.freshName("row")
+      val holder = ctx.freshName("holder")
+      val rowWriter = ctx.freshName("createRowWriter")
+      val arrayWriter = ctx.freshName("createArrayWriter")
+      val unsafeRowClass = classOf[UnsafeRow].getName
+      val unsafeArrayClass = classOf[UnsafeArrayData].getName
+      val holderClass = classOf[BufferHolder].getName
+      val rowWriterClass = classOf[UnsafeRowWriter].getName
+      val arrayWriterClass = classOf[UnsafeArrayWriter].getName
+      ctx.addMutableState(unsafeRowClass, row, "")
+      ctx.addMutableState(unsafeArrayClass, array, "")
+      ctx.addMutableState(holderClass, holder, "")
+      ctx.addMutableState(rowWriterClass, rowWriter, "")
+      ctx.addMutableState(arrayWriterClass, arrayWriter, "")
+      val unsafeArraySizeInBytes =
+        UnsafeArrayData.calculateHeaderPortionInBytes(size) +
+        ByteArrayMethods.roundNumberOfBytesToNearestWord(dt.defaultSize * size)
+
+      // To write data to UnsafeArrayData, we create UnsafeRow with a single array field
+      // and then prepare BufferHolder for the array.
+      // In summary, this does not use UnsafeRow and wastes some bits in an byte array
+      (s"""
+        $row = new $unsafeRowClass(1);
+        $holder = new $holderClass($row, $unsafeArraySizeInBytes);
+        $rowWriter = new $rowWriterClass($holder, 1);
```

Comment:
@cloud-fan, I could do it. One question. How about filling 0 when we call `UnsafeArrayData.setNullAt(ordinal)`? Since `UnsafeArrayDataWriter(ordinal)` has `setNull<Type>`, it can fill 0 to a column (`ordinal`) based on each `Type`. `UnsafeArrayData.setNullAt(ordinal)` does not know the type. Can we assume that 0 is already filled at a column (`ordinal`) when we call `UnsafeArrayData.setNullAt(ordinal)`?